### PR TITLE
Implement Job Management Package

### PR DESCRIPTION
**Pull request by rchammonds**
First PR for the challenge!

Implements a "job" package which handles the job management portion of the system. It exposes the API defined in the design doc, with the exception of `job.Stdout/Stderr` methods which return `(io.ReaderCloser, error)` instead of just a `io.ReadCloser` as those methods are fallible. 

You'll also find a `streamer` package, which provides a utility for reading from a file that has an active writer (a "live" file). This streamer implements the `io.ReadCloser` interface and automatically returns an io.EOF when the writer closes the file. I hand cranked my own inotify utility to implement this, since `fsnotify` could not return an event on file close.

As you've probably guessed, the `streamer` package handles most of the concerns around streaming stdout/stderr. The ability to detect a file close notification is extremely useful, as it means I don't have to propagate any sort of "done" notification from the job to the streamer. The `streamer.LiveFileStreamer` implementation is decoupled from the job package and stands on its own. 

Run tests with `make test` and/or `make test-race`.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+test:
+	@go test -v ./...
+
+# requires c compiler
+test-race: echo
```
Comment (zmb3) [inline Makefile:7]: `test` and `test-race` should also be `PHONY` targets, right?

File: testdata/testprograms/echo
Comment (zmb3) [inline testdata/testprograms/echo:1]: Mistakenly committed?

File: job/job.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+type Status struct {
+	CurrentState State
+	ReturnCode   *int
+}
+
+type JobArgs struct {
+	Command    string
+	Args       []string
+	StdoutPath string
+	StderrPath string
```
Comment (zmb3) [inline job/job.go:50]: Should these be exported? From my perspective, the fact that you're using files to power the output streaming functionality is an implementation detail and not something you want consumers of the package to rely on.

File: job/job.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+type Job struct {
+	sync.Mutex
```
Comment (zmb3) [inline job/job.go]: I don't love embedding the mutex like this, because you make the `Lock` and `Unlock` functions part of the job's public API. Another package that consumes your job library needs to only call `Lock` at the wrong time to get your library to deadlock.

File: job/job.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+		Args: args.Args,
+	}
+
+	// Create our output files!
+	stdoutFile, err := createOutputFile(args.StdoutPath)
+	stderrFile, err2 := createOutputFile(args.StderrPath)
+	if errors.Join(err, err2) != nil {
+		logFileClose(stdoutFile)
+		logFileClose(stderrFile)
+		return nil, fmt.Errorf("error creating output file(s): %w", err)
```
Comment (zmb3) [inline job/job.go:87]: Do you intentionally only wrap `err` and not `err2` here?

File: job/job_test.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	if !stdout {
+		prefix = "stderr"
+	}
+
+	bldr := strings.Builder{}
+	for idx := range count {
+		bldr.WriteString(prefix)
+		bldr.WriteString(" ")
+		bldr.WriteString(strconv.Itoa(idx + 1))
+		bldr.WriteString("\n")
```
Comment (zmb3) [inline job/job_test.go]: Alternatively:
```suggestion
		fmt.Fprintln(bldr, "%v %d", prefix, idx+1)
```

File: job/job.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+func logFileClose(f *os.File) {
+	if err := f.Close(); err != nil {
+		if !errors.Is(err, os.ErrInvalid) {
+			slog.Error(fmt.Sprintf("Failed to close file: %s", err))
+		}
+	}
+}
+
+func NewJob(args JobArgs) (*Job, error) {
```
Comment (zmb3) [inline job/job.go]: `job.NewJob` stutters a bit. How about `job.New` instead?

File: streamer/streamer.go
Comment (zmb3) [inline streamer/streamer.go:1]: No changes requested, but this might make sense as an `internal/` package to avoid leaking implementation details outside the library.

File: streamer/streamer.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		File:         file,
+		closeOnce:    &sync.Once{},
+	}
+}
+
+func (l *LiveFileStreamer) fileRead(p []byte) (int, error) {
+	count, err := l.File.Read(p)
+	if err != nil {
+		// Read returns 0, io.EOF, so we need not check count
+		// or deal with data in 'p' on EOF
```
Comment (zmb3) [inline streamer/streamer.go]: Are you sure? Godocs say this:

```
When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. It may return the (non-nil) error from the same call or return the error (and n == 0) from a subsequent call. An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF.

Callers should always process the n > 0 bytes returned before considering the error err. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors.
```

File: streamer/streamer.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+			l.done = err
+			return count, l.done
+		}
+	}
+	return count, nil
+}
+
+func (l *LiveFileStreamer) Read(p []byte) (int, error) {
+	if l.done != nil {
+		return 0, l.done
```
Comment (zmb3) [inline streamer/streamer.go]: Doesn't the `io.Reader` contract require that we return `io.EOF` when done?

File: streamer/streamer.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	if !l.useWatch {
+		return l.fileRead(p)
+	}
```
Comment (zmb3) [inline streamer/streamer.go]: I see what you're doing now, and it probably works, but I find it a bit hard to follow since the core logic is split across `fileRead` and `Read` and `fileRead` is leaving behind state to tell the next read how to behave.

I wonder if we can express the same logic in a simpler way where it's all in a single function and not split up like this.

File: job/job.go (hunk 2 above)
Comment (rchammonds) [reply job/job.go:50]: They should not. Looks like I got a little fast and loose with exported/unexported fields.

Actually, same with the `streamer.LiveFileStreamer`.  The `File` and `WriteWatcher` fields should probably be private since the implementation takes ownership of those objects after calling the constructor.

File: job/job.go (hunk 3 above)
Comment (rchammonds) [reply job/job.go]: Great point. I'll refactor the mutex to an unexported field.

File: job/job.go (hunk 4 above)
Comment (rchammonds) [reply job/job.go:87]: No 😩 . I should be grabbing the result of `errors.Join` and wrapping that.

File: streamer/streamer.go (hunk 7 above)
Comment (rchammonds) [reply streamer/streamer.go]: That's true for the more generic contract defined for an `io.Reader`. In this instance, we're reading from an `os.File` though, where the documentation for the `Read` method says:
`At end of file, Read returns 0, io.EOF.`

So the `os` package has more strict semantics for file reads.

File: streamer/streamer.go (hunk 8 above)
Comment (rchammonds) [reply streamer/streamer.go]: It does for the happy path, but implementations need some way to communicate read errors. I believe it's acceptable to return non-EOF errors when an implementation encounters an internal error.

File: streamer/streamer.go (hunk 9 above)
Comment (rchammonds) [reply streamer/streamer.go]: I'm happy to take a stab at refactoring it! Agreed that the logic is a bit unwieldy, and this function went through several iterations as I migrated from `fsnotify` to my own `FileWatcher`. Let me give it another shot.

File: streamer/streamer.go (hunk 7 above)
Comment (rchammonds) [reply streamer/streamer.go]: Though, I guess this could be a footgun for a future developer that changes this `os.File` to an `io.Reader` without realizing the differences.

Comment (zmb3) [reply streamer/streamer.go]: > So the os package has more strict semantics for file reads.

Thanks for pointing this out. I'm fine with leaving this as-is. It's a short project, so we're less concerned about footguns for future developers than we might otherwise be.

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+	CGO_ENABLED=1 go test -race ./...
```
Comment (rosstimothy) [inline Makefile:8]: Is CGO required for the race detector? Is there any reason to have two separate test targets?

File: job/job.go [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+		// ExitCode() returns -1 if the process hasn't exited or was terminated by a signal
+		// Well we know the process exited
+		if e != nil && e.ExitCode() == -1 {
+			status, ok := e.ProcessState.Sys().(syscall.WaitStatus)
+			if ok && status.Signal() == syscall.SIGKILL {
+				// I suppose we can't know for certain that it this program who sent
+				// the SIGKILL, but it was *probably* us
+				// Maybe the documentation around the "STOPPED" state should instead be
+				// that this state means the program was forcefully terminated, rather than
+				// terminated *by a user*
```
Comment (rosstimothy) [inline job/job.go]: If this transition was handled in a different layer, or was enriched with information from the gRPC service you could probably infer better that a process was stopped manually by a human.

File: job/job.go [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+	exitErr       *exec.ExitError
+
+	stdoutPath string
+	stderrPath string
+}
+
+func logFileClose(f *os.File) {
+	if err := f.Close(); err != nil {
+		if !errors.Is(err, os.ErrInvalid) {
+			slog.Error(fmt.Sprintf("Failed to close file: %s", err))
```
Comment (rosstimothy) [inline job/job.go]: Suggestion: embrace structured logging 
```suggestion
			slog.Error("Failed to close file", "error", err)
```

File: job/job.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+	// current user process can read/write, group members can read
+	return os.OpenFile(path, flags, 0640)
+}
+
+func (j *Job) Status() Status {
+	var exitErr *exec.ExitError
+	var exited bool
+	j.Lock()
+	exited = j.processExited
+	exitErr = j.exitErr
```
Comment (rosstimothy) [inline job/job.go]: Since this is a pointer are there any concerns with interacting with it after releasing the lock?

File: job/job.go [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+		CurrentState: NewState(exited, exitErr),
+		ReturnCode:   exitCode,
+	}
+}
+
+func (j *Job) Stop() error {
+	var err error
+	j.Lock()
+	if !j.processExited {
+		err = j.cmd.Process.Kill()
```
Comment (rosstimothy) [inline job/job.go:166]: Should the lock be held while issuing the kill signal? Perhaps rewriting this to exit early if the process already exited would clean up some of the error handling here as well?

File: job/job.go [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+		} // Drain watcher events
+	}
+
+	if err != nil {
+		// err has already been wrapped with contextual message
+		return nil, err
+	}
+
+	return streamer.NewLiveFileStreamer(readHandle, watcher), nil
+}
```
Comment (rosstimothy) [inline job/job.go]: Are there any changes that could be made to the streamer package to make the API easier to use for callers? There is a lot of sublty and nuance going on in this function to accomodate the streamer package.

File: Makefile (hunk 10 above)
Comment (rchammonds) [reply Makefile:8]: Unfortunately, yes. And that's why I broke it up into two separate targets. If you're on a system that doesn't a C compiler ready to go you can at least still run the tests without the race detector.

File: job/job.go (hunk 13 above)
Comment (rchammonds) [reply job/job.go]: Ah.. yes there is and I should have just made a copy of `exitErr` to return to the caller. I'll get this fixed.

File: job/job.go (hunk 14 above)
Comment (rchammonds) [reply job/job.go:166]: I think we definitely want to hold the lock while doing this especially considering the server implementation may receive multiple concurrent requests to stop the job. The documentation doesn't explicitly say that `Process.Kill` is safe for concurrent calls, which usually means it is not. I also don't see any warnings in the documentation about this function potentially blocking which leads me to believe that we won't be holding up anyone else for too terribly long.

File: job/job.go (hunk 15 above)
Comment (rchammonds) [reply job/job.go]: I had to think on it for a few minutes, but yea, I have an idea.

I could write a constructor `streamer.NewLiveFileStreamerFromPath(path string)` that handles creation of the read handle and watcher internally. I'll also add a `StopWatching` method to close the watcher owned by the `LiveFileStreamer`. 

The order of operations becomes pretty simple:
- Create the LiveFileStreamer
- Lock the job and inspect exit status
- if exited, stop the watcher, otherwise do nothing (we don't need to hold the lock while calling `StopWatching``)

The internal `Watcher` is completely hidden from the caller, which is probably how it should've been from the start. If that sounds alright with you, I'll implement this when I get some time this afternoon.

File: job/job.go (hunk 2 above)
Comment (rchammonds) [reply job/job.go:50]: I misunderstood earlier.  I'm intentionally asking the caller to specify these paths because they might be interested in controlling where the process output ends up. 

For example, my gRPC server will use this package with the expectation that no job data will persist beyond a restart of the server, let alone a reboot of the machine, so it directs output to `/tmp`. Maybe another consumer wants that ensure that the output data survives a reboot. I think the flexibility could be useful.

File: job/job.go (hunk 15 above)
Comment (rosstimothy) [reply job/job.go]: I like making the watcher an internal implementation detail, though I might suggest going one step further. Can the job provide the streamer package with the path _and_ wether the job is still running? Then everything related to the watcher can remain internal.

Comment (rchammonds) [reply job/job.go]: It's easy enough to give the LiveFileStreamer/Watcher a channel to listen on to determine if the process (and therefore the writer) has exited. I think I just got tunnel vision on having the streamer/watcher stop itself when the writer closes the file. 

Listening for the job to simply close a channel is easy enough. I didn't need to roll my own watcher implementation on top of inotify then, but that's alright. It was an interesting learning experience.

Comment (rchammonds) [reply job/job.go]:
```
func (j *Job) watchOutput(path string) (io.ReadCloser, error) {
	fileStreamer, err := streamer.NewLiveFileStreamer(path, j.processDone)
	if err != nil {
		return nil, fmt.Errorf("failed to create file streamer: %w", err)
	}
	return fileStreamer, nil
}
```

That's quite a bit simpler

File: streamer/streamer.go (hunk 9 above)
Comment (rchammonds) [reply streamer/streamer.go]: Cleaned this up quite a bit with my last commit. Let me know what you think.

File: job/job.go (hunk 2 above)
Comment (zmb3) [reply job/job.go:50]: Ok, we can leave it as-is for the purpose of this project, but I do think it's leaking implementation details.

The core function that the library is meant to expose is the ability to stream output. Ideally, you could completely change the implementation under the hood (say from writing to files to writing to an in-memory buffer) without the client knowing. That's hard to do once your public API exposes the fact that we're writing to files.

File: streamer/streamer_test.go [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+func (t *TestResources) Cleanup() {
+	_ = t.WriteHandle.Close()
+}
+
+func NewTestResources(dir string) (TestResources, error) {
+	var allErrs error
+	testFile, err := os.OpenFile(filepath.Join(dir, "test"), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0666)
+	allErrs = errors.Join(allErrs, err)
```
Comment (zmb3) [inline streamer/streamer_test.go:144]: What's going on with the `errors.Join` here. Won't you have at most 1 error here on any invocation?

File: streamer/streamer.go (hunk 9 above)
Comment (zmb3) [reply streamer/streamer.go]: Looks much better, thanks for the update.

File: job/job.go [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+				// The process has exited, it was killed by a kill signal,
+				// and the job has a 'userKilled' flag indicating that the caller
+				// asked us to stop a running job.
+				return JobStatusStopped
+			}
+		}
+		return JobstatusComplete
+	}
+	return JobStatusRunning
+}
```
Comment (rosstimothy) [inline job/job.go]: Can this be simplified down to the following?
```suggestion
	if !processExited {
		return JobStatusRunning
	}
	
	if userKilled {
		return JobStatusStopped
	}
	
	return JobstatusComplete
}
```

File: job/job.go (hunk 2 above)
Comment (rchammonds) [reply job/job.go:50]: I think you've swung me. I'm coming at it from the perspective that a caller might want to direct stdout/stderr to different locations depending on the job... But if that's what the caller wants to do, they can implement it themselves by calling `job.Stdout` and directing the stream wherever they please. 

We could also extend the job API to include `job.ClearOutput` method to clean up any memory/files related to the job. That way the caller can free up any resources related to output management (files and/or memory) without knowing the implementation details.

File: job/job.go [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+	stdoutPath string
+	stderrPath string
+}
+
+func logFileClose(f *os.File) {
+	if err := f.Close(); err != nil {
+		if !errors.Is(err, os.ErrInvalid) {
+			slog.Error("Failed to close file", "error", err)
+		}
+	}
```
Comment (rosstimothy) [inline job/job.go]: If `ErrInvalid` is only returned when the file is nil perhaps return early in that case and drop the special error handling?
```suggestion
	if f == nil {
		return
	}
	
	if err := f.Close(); err != nil {
		slog.Error("Failed to close file", "error", err)
	}
```

File: job/job.go [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+		// We may consider holding the lock until the files are
+		// closed, but I don't believe we need that guarantee
+		// Other methods can be assure that observing
+		// 'processExited == true' means that the last write to
+		// the output files have completed
+		defer newJob.jobLock.Unlock()
+
+		close(newJob.processDone)
+		newJob.processExited = true
+		newJob.exitErr, _ = err.(*exec.ExitError)
```
Comment (rosstimothy) [inline job/job.go]: Suggestion: use errors.As?

File: job/job.go [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+	}()
+
+	return newJob, err
+}
+
+func createOutputFile(path string) (*os.File, error) {
+	// We need to open a file for writing, create if not exists,
+	// and truncate existing files
+	flags := os.O_CREATE | os.O_WRONLY | os.O_TRUNC
```
Comment (rosstimothy) [inline job/job.go]:
```suggestion
	const flags = os.O_CREATE | os.O_WRONLY | os.O_TRUNC
```

File: streamer/streamer.go (hunk 8 above)
Comment (zmb3) [reply streamer/streamer.go]: Right, but you want to try to stick to the semantics of `io.Reader` as much as possible so that Go's io just works.

For example, the docs for `io.Copy` say:

>A successful Copy returns err == nil, not err == EOF. Because Copy is defined to
> read from src until EOF, it does not treat an EOF from Read as an error to be reported.

If your streamer doesn't return `io.EOF` when it reaches the end of the stream, then `io.Copy` will return an error instead of succeeding, which can have consequences as an unexpected error bubbles up the stack.

One way to induce this behavior is to mimic a job that never writes any output. We can create a `LiveFileStreamer` pointed at an empty file, close it, and then try to `io.Copy` from it. The expected behavior would be that 0 bytes are copied and no error is returned, but with the current implementation we'll get an error.

File: job/job.go (hunk 15 above)
Comment (rosstimothy) [reply job/job.go]: Much simpler 👍!

File: streamer/streamer_test.go (hunk 16 above)
Comment (rchammonds) [reply streamer/streamer_test.go:144]: Yes. Previously I was initializing a couple other resources in this utility and so the join had purpose. I mistakenly left this over after refactoring last night. I'll get this cleaned up.

File: job/job.go (hunk 17 above)
Comment (rchammonds) [reply job/job.go]: That would probably be fine. I had assumed (maybe incorrectly) that there was a possibility that the process could still exit gracefully after the call to `process.Kill`. And for that reason, I think there might be value in doing our due diligence to inspect the process state and determine if it was indeed stopped via a kill signal or had a graceful exit. 

It's a very tight window, and likely won't matter much to the user. I'm open to simplifying to this suggestion.

File: streamer/streamer.go (hunk 8 above)
Comment (rchammonds) [reply streamer/streamer.go]: I've pushed some whitebox tests for the `streamer` package demonstrating that `LiveFileStreamer` is (to the best of my knowledge) meeting the `io.Reader` contract. One of the test uses `io.Copy` to copy data from the streamer into a buffered writer and shows that subsequent calls return `0, io.EOF`.

I'm also using other `io` utilities like `ReadAll` and `ReadFull` in a handful of tests for both the streamer and job package.
Am I still missing an edge case? I'll take a closer look after work this afternoon, but I'm having trouble spotting anything.

**Review by zmb3 — APPROVED** [review]

File: job/job.go (hunk 17 above)
Comment (rchammonds) [reply job/job.go]: Implemented

**Review by rosstimothy — APPROVED** [review]

File: streamer/streamer.go [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+
+	// A FileWriteWatcher that is watching the associated
+	// File handle
+	writeWatcher *FileWriteWatcher
+
+	// Indicates that the file will receive no more writes
+	writerDone chan struct{}
+
+	// manage close behavior
+	closeOnce *sync.Once
```
Comment (nklaassen) [inline streamer/streamer.go:36]: nit: i don't think this needs to be a pointer, if it's a plain value of type sync.Once there's no need to initialize it.

File: streamer/streamer_whitebox_test.go [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReadError(t *testing.T) {
+	writeHandle, err := os.CreateTemp(t.TempDir(), "")
+	require.NoError(t, err)
+	defer writeHandle.Close()
+	//os.OpenFile(filepath.Join(t.TempDir(),
```
Comment (nklaassen) [inline streamer/streamer_whitebox_test.go]:
```suggestion
```

File: streamer/streamer_whitebox_test.go [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+	fs, err := NewLiveFileStreamer(writeHandle.Name(), done)
+	require.NoError(t, err)
+
+	// Close the read handle in the middle of reading!
+	fs.file.Close()
+
+	// Reading should fail with a non-EOF error
+	_, err = io.ReadAll(fs)
+	require.Error(t, err)
+	require.False(t, errors.Is(err, io.EOF))
```
Comment (nklaassen) [inline streamer/streamer_whitebox_test.go:29]: what if i told you `require.NotErrorIs(t, err, io.EOF)` exists

**Review by nklaassen — APPROVED** [review]

File: streamer/streamer_whitebox_test.go (hunk 23 above)
Comment (rchammonds) [reply streamer/streamer_whitebox_test.go:29]: I would start using it, that's what!
